python39Packages.bsddb3,python312Packages.berkeleydb, gramps, exaile: mark broken from Pyt…

AkechiShiro opened the pull request:
…hon36, init 18.1.8, add berkeleydb dependency, add berkeleydb dependency

## Description of changes

Deprecate the use of bsddb3 from Python 3.6 and newer, it was replaced upstream by `berkeleydb`, see : https://www.jcea.es/programacion/pybsddb.htm

Berkeleydb was thus packaged and all occurences of `bsddb3` were replaced by `berkeleydb`.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
  </ul>
</details>

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
  </ul>
</details>

TomaSajt commented on pkgs/development/python-modules/berkeleydb/default.nix:
```diff
@@ -0,0 +1,45 @@
+    ${python.interpreter} test.py
+  '';
+
+  # Path to database need to be set.
+  # Somehow the setup.py flag is not propagated.
+  #setupPyBuildFlags = [ "--berkeley-db=${pkgs.db}" ];
+  # We can also use a variable
+  preConfigure = ''
+    export BERKELEYDB_DIR=${pkgs.db.dev};
+  '';
```
Sadly, `setupPyBuildFlags` doesn't work when using `pyproject = true`

Instead of doing `export SOMEVAR=value` inside a phase you can add `env.SOMEVAR = value;` into the derivation to achieve the same effect.

Also, I think you should set `BERKELEYDB_INCDIR` instead of just `BERKELEY_DIR`, becuase `_DIR` also expects the library files to be in the directory. The only reason it works is becuase `db` is in `buildInputs` too.

---

So, I think you should either keep `buildInputs` and do
```
env.BERKELEYDB_INCDIR = "${lib.getDev pkgs.db}/include";
```
or remove `buildInputs` and do
```
env = {
  BERKELEYDB_INCDIR = "${lib.getDev pkgs.db}/include";
  BERKELEYDB_LIBDIR = "${lib.getLib pkgs.db}/lib";
};
```

AkechiShiro commented:
> Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 6 packages failed to build:
> 15 packages built:

Pymilter will be fixed once this issue is fixed upstream : https://github.com/sdgathman/pymilter/issues/61

pinpox commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
  </ul>
</details>

ryneeverett commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -23,7 +23,15 @@ buildPythonPackage rec {
   buildInputs = [ pkgs.db ];
 
-  doCheck = pythonOlder "3.12"; # distutils usage
+  broken = pythonOlder "3.6";
```
By my reading it's not actually broken until 3.10

AkechiShiro replied:
Maybe yeah, I checked again and it is recommended to replace the dependency if possible for Python 3.6+ : 
![image](https://github.com/NixOS/nixpkgs/assets/14914796/2a574c5d-708f-4c7a-b47e-70cffb4f2436)

In the case, where we want to maximize compatibility, it's most likely a good idea to keep it until Python 3.10.

![image](https://github.com/NixOS/nixpkgs/assets/14914796/ad9e470f-0751-4637-a7a1-a50506a29497)

So you're right, I will update this.

ryneeverett commented:
The package I maintain -- exaile -- has supported berkeleydb for a long time now so this change is good. However, for others which don't yet support berkeleydb, wouldn't it be better to lock the input to python39 than to change to an input which will likely break them?

AkechiShiro commented:
Totally agree @ryneeverett

ryneeverett commented:
> Do you have any idea how I can lock maybe overriding ?

I believe it should just be a matter of changing the input from `python` to `python39`.

AkechiShiro commented:
@TomaSajt do you have any idea if gramps can be used with `berkeleydb` or we should keep the old `bsddb3`, meaning we would probably mark as `broken` gramps for `PythonAtLeast` 3.12

AkechiShiro commented:
@ryneeverett I don't see much package using `bsddb3` beside `gramps` and your package right now, I'm not sure locking it might be necessary for just 2 packages, but we can keep using it in case `berkeleydb` breaks `gramps` I guess.

AkechiShiro commented:
Also locking the input in the 2 pkgs to Python 3.9 would probably be an issue as if you would build other packages (with Python312), then I don't know if that would still work.

The best would be to use `berkeleydb` if Python >=3.9 else use `bsddb3` but if we can ditch `bsddb3` fully, then this might be better and less overhead for maintaining pkgs.

ryneeverett commented:
I don't want you to lock my package because exaile already does support berkeleydb.

My suggestion was to change the input for each individual package. For example, I think you could change the gramps input from `pythonPackages` to `python39Packages`.

AkechiShiro commented:
@ryneeverett it was not confirmed that `gramps` does not support `berkeleydb`, I will not lock your package if you wish to keep using `bsddb3`, you surely can, I can remove the new package introduced, but the package will never build on Python 3.12, I don't know if it's a problem or could become a future problem later.

AkechiShiro commented:
@ryneeverett I'll let you handle the update on your package as to keep both `bsddb3` on pythonOlder 3.9 and replace it with `berkeleydb` on Python 3.12, it will take more code to have this logic.

I'm undoing the change for the exaile pkg.

ryneeverett commented:
I'm not sure what's going wrong with the communication here. We seem to have a different understanding of "locking" a dependency and perhaps that terminology isn't helpful.

Exaile *does* support berkeleydb. Your initial change was unproblematic for exaile. It was the other packages I was expressing concern for.

> @ryneeverett it was not confirmed that `gramps` does not support `berkeleydb`

I think you should err on the side of not breaking things. However, there are few enough packages here that it's easy enough to check each one. A quick search indicates that gramps supports both bsdd3 and berkeleydb: https://github.com/search?q=repo%3Agramps-project%2Fgramps%20berkeleydb&type=code

AkechiShiro commented:
Thanks, but yeah we need to be sure what locking means, so far I've understood that we want to keep using `bsdbb3` if we're using Python 3.9 or older.

And if we're using much recent Python versions, we should prefer `berkeleydb` if supported, else if the package doesn't support `berkeleydb`, we probably mark it as broken for recent Python version (3.10+)

AkechiShiro committed:
python312Packages.berkeleydb: init at 18.1.8

* Add env fix from tomaSajt
* nixformat
* better format for imports
* remove unused import

AkechiShiro committed:
exaile: replace bsddb3 with berkeleydb

Prepare for Python 3.12 compatibility as bsddb3 is not officially
compatible with Python 3.10+ and Python 3.12 breaks bsddb3 tests due to
distutils being deprecated.

AkechiShiro committed:
gramps: replace bsddb3 with berkeleydb

Prepare for Python 3.12 compatibility as bsddb3 is not officially
compatible with Python 3.10+ and Python 3.12 breaks bsddb3 tests due to
distutils being deprecated

AkechiShiro commented:
@ryneeverett do you agree with the latest changes ? 

I'm sorry also that there was a misunderstanding, I hope this is cleared up.

Basically, I believe that if an old release of exaile or gramps is overriden by some other, they could need `bsddb3`, this should help be to err on the safe side.

TomaSajt commented on pkgs/applications/misc/gramps/default.nix:
```diff
@@ -69,11 +69,13 @@ buildPythonApplication rec {
 
   propagatedBuildInputs = with pythonPackages; [
-    bsddb3
+    berkeleydb
```
This should probably be removed

TomaSajt commented on pkgs/applications/audio/exaile/default.nix:
```diff
@@ -51,13 +51,14 @@ stdenv.mkDerivation rec {
     gst-python
-  ]) ++ lib.optional deviceDetectionSupport udisks
+  ] ++ lib.optional (pythonAtLeast "3.10") berkeleydb
+  ++ lib.optional (pythonOlder "3.9") bsddb3)
```
`pythonOlder` is `<` not `<=`, so if you had Python 3.9, you wouldn't have either of the two packages.
It should probably be `pythonOlder "3.10"`

TomaSajt commented on pkgs/applications/misc/gramps/default.nix:
```diff
@@ -69,11 +69,13 @@ buildPythonApplication rec {
     pycairo
-  ];
+  ] ++ lib.optional (pythonAtLeast "3.10") berkeleydb
+  ++ lib.optional (pythonOlder "3.9") bsddb3;
```
Same as with exaile

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -21,8 +21,16 @@ buildPythonPackage rec {
-  doCheck = pythonOlder "3.12"; # distutils usage
+  # See : https://github.com/NixOS/nixpkgs/pull/311198#discussion_r1599257522
+  # More details here : https://www.jcea.es/programacion/pybsddb.htm
+  broken = pythonAtLeast "3.9" || pythonOlder "3.5";
```
AFAICT Python 3.9 is supported, so this should be `pythonAtLeast "3.10"

TomaSajt replied:
`pythonOlder "3.5"` should be removed, since the link you provided specifically says that only this package of the two can support python <= 3.5

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -31,7 +39,13 @@ buildPythonPackage rec {
   #setupPyBuildFlags = [ "--berkeley-db=${pkgs.db}" ];
+
   # We can also use a variable
+  # See : https://github.com/NixOS/nixpkgs/pull/311198#discussion_r1597746759
+  env = {
+    BERKELEYDB_INCDIR = "${lib.getDev pkgs.db}/include";
+    BERKELEYDB_LIBDIR = "${lib.getLib pkgs.db}/lib";
+  };
```
I originally meant this to go to the other package (see where I made the comment).
Now I see that both packages have this structure.

I think we should just drop this change for now, maybe in another PR we can add it back.

TomaSajt replied:
Also, the reason why it wasn't working when you removed `buildInputs` was because you forgot to remove the `preConfigure` below these lines.

AkechiShiro replied:
Seems I did something wrong, thanks for the review, I'll fix this.

AkechiShiro replied:
@TomaSajt  should I remove from `berkeleydb` the change and make it in another PR ?

TomaSajt replied:
I think you should remove it, but you don't need to create another PR, it's not that important

AkechiShiro committed:
python39Packages.bsddb3: disable on Python 3.10+, remove distutils usage

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages failed to build:</summary>
  <ul>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
  </ul>
</details>

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages failed to build:</summary>
  <ul>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
  </ul>
</details>

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>

AkechiShiro commented:
@TomaSajt if you can review again one last time this PR ?

ryneeverett reviewed (commented):
I don't see the purpose of pythonAtLeast/pythonOlder when both exaile and gramps support berkeleydb. You can simply switch the input from bsddb3 to berkeleydb for those packages, as you did in the original commit.

AkechiShiro commented:
@ryneeverett I thought it'd be useful to keep `bsddb3` if anyone would build an older release of `gramps` or `exaile` that doesn't support `berkeleydb` (`exaile` support it since the 6 December 2021) but I'm not sure if anyone would want to do that or if that would make 

I think you're probably right, it would be better to remove `bsddb3`, I can't really find a use case for building a very very old release for those packages.

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python311Packages.bsddb3</li>
    <li>python311Packages.bsddb3.dist</li>
    <li>python311Packages.pymilter</li>
    <li>python311Packages.pymilter.dist</li>
    <li>python311Packages.scrapy-deltafetch</li>
    <li>python311Packages.scrapy-deltafetch.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
    <li>python312Packages.bsddb3</li>
    <li>python312Packages.bsddb3.dist</li>
    <li>python312Packages.pymilter</li>
    <li>python312Packages.pymilter.dist</li>
    <li>python312Packages.scrapy-deltafetch</li>
    <li>python312Packages.scrapy-deltafetch.dist</li>
    <li>spf-engine</li>
    <li>spf-engine.dist</li>
  </ul>
</details>

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -1,13 +1,14 @@
 }:
 
+let
+  inherit (pythonPackages) buildPythonPackage pythonOlder pythonAtLeast python;
```
I think you should revert this change as it was before

TomaSajt commented on pkgs/applications/misc/gramps/default.nix:
```diff
@@ -1,26 +1,27 @@
 
 let
-  inherit (pythonPackages) buildPythonApplication pythonOlder;
+  inherit (pythonPackages) buildPythonApplication pythonOlder pythonAtLeast;
```
```suggestion
  inherit (pythonPackages) buildPythonApplication pythonOlder;
```

Not used anymore

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -21,12 +22,9 @@ buildPythonPackage rec {
-  '';
+  # See : https://github.com/NixOS/nixpkgs/pull/311198#discussion_r1599257522
+  # More details here : https://www.jcea.es/programacion/pybsddb.htm
+  broken = pythonAtLeast "3.10";
```
```suggestion
  disabled = pythonAtLeast "3.10";
```

`broken` doesn't seem to do anything...
If you use `disabled` it will not even try to build it.

TomaSajt replied:
Or do you want to mark it as broken, but still let people use it if they allow broken packages?

AkechiShiro replied:
Yes I believe it would be better to give a choice, only allowing broken package can be done by defining `NIXPKGS_ALLOW_BROKEN=1`

SuperSandro2000 replied:
We use disabled for those cases where we know the package will not work.

TomaSajt commented on pkgs/applications/misc/gramps/default.nix:
```diff
@@ -1,26 +1,27 @@
+  pythonPackages,
+  glibcLocales,
+  intltool,
+  gexiv2,
+  pango,
+  gobject-introspection,
+  wrapGAppsHook3,
+  gettext,
   # Optional packages:
-, enableOSM ? true
-, osm-gps-map
-, glib-networking
-, enableGraphviz ? true
-, graphviz
-, enableGhostscript ? true
-, ghostscript
+  enableOSM ? true,
+  osm-gps-map,
+  glib-networking,
+  enableGraphviz ? true,
+  graphviz,
+  enableGhostscript ? true,
+  ghostscript,
```
Why the change to the formatting?

If you really want to use this formatting, use `nixfmt-rfc-style` on the entire file, but I think this is unnecessary for this PR.

This suggestion also applies to all other files that had formatting changes.

AkechiShiro replied:
Thanks will do so, I wasn't aware of the tool name and `nixfmt` currently does not do a good job I believe.

SuperSandro2000 replied:
I would recommend to just drop this commit form this PR

AkechiShiro replied:
I don't understand why or the reasoning behind your comment @SuperSandro2000

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -36,6 +34,17 @@ buildPythonPackage rec {
     export BERKELEYDB_DIR=${pkgs.db.dev};
   '';
 
+  postPatch = ''
+    substituteInPlace test3.py \
+      --replace-fail "from distutils.util import get_platform" "import platform" \
+      --replace-fail "get_platform()" "f'{platform.system().lower()}-{platform.machine()}'" \
+      --replace-fail "_TextTestResult" "TextTestResult"
+  '';
```
Since we're no longer allowing python >=3.10 this no longer seems necessary.

AkechiShiro replied:
Shouldn't we leave this in case a new package is added in Nixpkgs that do depend on `bsddb3` but does not support `berkeleydb` ? I don't see a reason, why it should be removed, right now, it doesn't break anything to leave it.

TomaSajt replied:
If the plan is supporting `>=3.10` then keep it, sure.

SuperSandro2000 commented on pkgs/development/python-modules/berkeleydb/default.nix:
```diff
@@ -0,0 +1,43 @@
+    description = "Python bindings for Oracle Berkeley DB";
+    homepage = "https://www.jcea.es/programacion/pybsddb.htm";
+    license = with licenses; [ bsd3 ];
+    maintainers = [ ];
```
We should set this

AkechiShiro replied:
I'm not able to commit to maintain this package sadly.

AkechiShiro replied:
I will update this sorry, I forgot to do so.

AkechiShiro replied:
I did check there is nothing to update, see : https://pypi.org/project/berkeleydb/

SuperSandro2000 commented on pkgs/development/python-modules/berkeleydb/default.nix:
```diff
@@ -21,8 +22,6 @@ buildPythonPackage rec {
     setuptools
   ];
 
-  buildInputs = [ pkgs.db ];
```
That commit accidentally removed the buildInput

AkechiShiro replied:
It was intended to be removed as @TomaSajt did recommend previously, please read the history of this PR.

SuperSandro2000 commented on pkgs/applications/misc/gramps/default.nix:
```diff
@@ -69,13 +69,14 @@ buildPythonApplication rec {
     pycairo
   ];
 
 
+
```
```suggestion
```

TomaSajt replied:
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/bsddb3/default.nix:
same for this commit

SuperSandro2000 reviewed (commented):
The history needs some cleanup. Also the commits should use python3Packages

AkechiShiro commented:
@SuperSandro2000 @TomaSajt could you please review one more time, sorry I left the PR in a broken state last time

TomaSajt commented:
@AkechiShiro
Before we continue this PR, could you please run `git rebase --interactive HEAD~8` and squash the related commits together? *(8 is the number of commits we're working with)*


```gitrebase
pick 8249698c6459 python3Packages.berkeleydb: init at 18.1.8
fixup 75fd18b275d1 python3Packages.berkeleydb: nixformat
pick a3b54f874865 exaile: replace bsddb3 by berkeleydb
pick eae86203b227 gramps: remove bsddb3, replace by berkeleydb
fixup 60192f6a43b9 python3Packages.gramps: remove unused import
drop 2304365f89e3 gramps: better format for imports
pick 1d5c4f87e413 python3Packages.bsddb3: fix test failure for Python 3.12+
fixup d8fedfbcf056 python3Packages.bsddb3: mark disabled for Python 3.10+
```

(Notice the first words are changed from `pick` to `fixup` in some cases. In those cases the changes of the commit marked as `fixup` will be squashed together with the commit above)
(I also marked the `gramps` formatting commit as `drop`, since it's not needed)

I also recommend changing `pick` to `reword` so that you can change the commit messages for the squashed commits, if you want to.

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -1,14 +1,14 @@
 }:
 
-buildPythonPackage rec {
+python3Packages.buildPythonPackage rec {
```
```suggestion
buildPythonPackage rec {
```
*(And also do the relevant change in the top attrset)*

`python3Packages` only ever has to be used when dealing with packages outside of the `python-modules` directory. Inside the directory, the top attrset's values are also populated by `python3Packages`'s values.

*Also, even if you wanted to do this, you'd have to use `pythonPackages`, since, it turns out `python3Packages` is always an alias to `python311Packages`.*

AkechiShiro replied:
Thanks for the explanation, I will keep this in mind for future PRs.

AkechiShiro replied:
One check seems to fail after having removed the `python3Packages` : https://github.com/NixOS/nixpkgs/actions/runs/9528874875/job/26267015309?pr=311198

AkechiShiro replied:
I have no idea why it is failing only for `bsddb3`, for `berkeleydb`, it does work as you described

TomaSajt commented on pkgs/development/python-modules/berkeleydb/default.nix:
```diff
@@ -0,0 +1,39 @@
+  setuptools,
+  pkgs,
+  python,
+  pythonOlder,
```
```suggestion
```

this was unused

AkechiShiro replied:
Done

AkechiShiro commented:
I've done a mistake on the rebase, will fix it as soon as possible

TomaSajt reviewed (changes requested):
Please change the commit messages (and the PR title also) to use `python39Packages.bsddb3` and `python312Packages.berkeleydb`.

~~It looks like `pythonPackages` (without the number) is for `python2`. This might be why `buildPythonPackage` was not defined in the CI, since I guess it doesn't work on `python2` and it was trying to use it.~~ <--- This is wrong, sorry

Not sure why the CI was failing...

*Just using `python3Packages` is not really acceptable, since IIRC the CI isn't allowed to use aliases*

AkechiShiro commented:
> Please change the commit messages (and the PR title also) to use `python39Packages.bsddb3` and `python312Packages.berkeleydb`.
> 
> Not sure why the CI was failing...

It was also failing locally on my side.

Will change the commits yes

TomaSajt commented:
> It was also failing locally on my side.

I guess I just wasn't clear enough on my previous change request, since I couldn't use GitHub's builtin tool to request two line changes.
This is what I meant:
```diff
diff --git a/pkgs/development/python-modules/bsddb3/default.nix b/pkgs/development/python-modules/bsddb3/default.nix
index bb911e28951d..d91836bd9dea 100644
--- a/pkgs/development/python-modules/bsddb3/default.nix
+++ b/pkgs/development/python-modules/bsddb3/default.nix
@@ -2,13 +2,13 @@
   lib,
   pythonAtLeast,
   python,
-  pythonPackages,
+  buildPythonPackage,
   fetchPypi,
   setuptools,
   pkgs,
 }:

-pythonPackages.buildPythonPackage rec {
+buildPythonPackage rec {
   pname = "bsddb3";
   version = "6.2.9";
   pyproject = true;
```

AkechiShiro commented:
I think the PR is ready for review @TomaSajt

Not sure if there is still stuff to fix for the merge

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -36,6 +36,17 @@ buildPythonPackage rec {
 
+  postPatch = ''
+    substituteInPlace test3.py \
+      --replace-fail "from distutils.util import get_platform" "import platform" \
+      --replace-fail "get_platform()" "f'{platform.system().lower()}-{platform.machine()}'" \
```
I just did some detective work and it looks like the testing code broke around 10000 commits ago, because the build directory structure generated by python changed, and since then `checkPhase` doesn't do anything, it only prints: "Ran 0 tests"

The following changes re-enables testing by using the proper imports:
```suggestion
      --replace-fail "from distutils.util import get_platform" "from sysconfig import get_platform" \
      --replace-fail "sys.version[0:3]" "sys.implementation.cache_tag" \
```

AkechiShiro replied:
Thanks a lot for your investigation of the root problem and this proposed fix.

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -36,6 +36,17 @@ buildPythonPackage rec {
+    substituteInPlace test3.py \
+      --replace-fail "from distutils.util import get_platform" "import platform" \
+      --replace-fail "get_platform()" "f'{platform.system().lower()}-{platform.machine()}'" \
+      --replace-fail "_TextTestResult" "TextTestResult"
```
I no longer think it's a good idea to try to support python >=3.10.
- upstream explicitly says that it's not supported by them
- the testing code doesn't work on later python versions
- we're using `disabled = pythonAtLeast "3.10";`, so it would always require overriding the package

So if we don't care about >=3.10, we can drop this line:
```suggestion
```

AkechiShiro replied:
We could remove all the block of subsitute and replace-fail too no ?

TomaSajt replied:
That other part was broken anyways, as it was not actually running any tests.

AkechiShiro replied:
Is it all good now ?

TomaSajt commented on pkgs/development/python-modules/bsddb3/default.nix:
```diff
@@ -21,12 +21,12 @@ buildPythonPackage rec {
-  checkPhase = ''
-    ${python.interpreter} test.py
-  '';
+  # marked disabled not broken following comments on PR
+  # https://github.com/NixOS/nixpkgs/pull/311198#discussion_r1623663114
```
```suggestion
```

IMO this comment should be removed

AkechiShiro commented:
Result of `nixpkgs-review pr 311198` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>exaile</li>
    <li>gramps</li>
    <li>gramps.dist</li>
    <li>python311Packages.berkeleydb</li>
    <li>python311Packages.berkeleydb.dist</li>
    <li>python312Packages.berkeleydb</li>
    <li>python312Packages.berkeleydb.dist</li>
  </ul>
</details>

SuperSandro2000 reviewed (approved):
I've cleaned up the commit messages a bit to speed things up
